SD-1436 split cells into cards

cryogenian opened the pull request:
<img width="1280" alt="2016-03-28 13 48 05" src="https://cloud.githubusercontent.com/assets/10245930/14076616/0dcb7520-f4ec-11e5-8f16-874f67833c99.png">
<img width="1280" alt="2016-03-28 13 48 23" src="https://cloud.githubusercontent.com/assets/10245930/14076617/0dd2eff8-f4ec-11e5-8701-d4379e22c2d9.png">
<img width="1280" alt="2016-03-28 13 48 33" src="https://cloud.githubusercontent.com/assets/10245930/14076618/0ded9d3a-f4ec-11e5-8c49-77c239bbe94e.png">
<img width="1280" alt="2016-03-28 13 48 40" src="https://cloud.githubusercontent.com/assets/10245930/14076619/0df55296-f4ec-11e5-9d1a-f04a30d84a50.png">
<img width="1280" alt="2016-03-28 13 49 10" src="https://cloud.githubusercontent.com/assets/10245930/14076620/0e12d5e6-f4ec-11e5-8788-28014ae03068.png">
<img width="1280" alt="2016-03-28 13 49 19" src="https://cloud.githubusercontent.com/assets/10245930/14076621/0e1adde0-f4ec-11e5-9100-b4aab89eb734.png">
<img width="1276" alt="2016-03-28 13 49 43" src="https://cloud.githubusercontent.com/assets/10245930/14076622/0e387684-f4ec-11e5-9e9c-022682a98a3b.png">
<img width="1279" alt="2016-03-28 13 49 57" src="https://cloud.githubusercontent.com/assets/10245930/14076623/0e449ba8-f4ec-11e5-8525-05c47bfc7f1f.png">

This is almost done, what do you think about it @jdegoes, @garyb, @jonsterling, @beckyconning 

+ All cells are identical now. I've removed every mark of being editor or results cell part.
+ Caching icon is removed. We're going to add caching card. 
+ Part of rendering logic was moved from `CellDef` to `CellType`. 
+ All cells autorun. We have no caching now, so, it's not a problem. _Play_ button changed to _Refresh_ although its logic hasn't been changed. Probably this button should be deleted with those message panels.
+ Unicode
+ Cells/cards show parents' error on yellow background. I think that this message panels will be fused to error card/next-action-error card. And exact message to backside of card. So this collapsing behaviour and yellow color is temp solution. 
+ Collapsed cards having errored parent can't be removed and have no messages.


Need to do 
+ Show all 10 card type in next action card. 
+ Tests.

beckyconning commented:
Looks good, why does the last one have the header "Download" it looks like "Visualisation options" followed by "Visualisation" to me?

cryogenian commented:
@beckyconning Thanks for noticing this! I'll fix it.

jdegoes commented:
This is an intermediate step, so I don't have much feedback.

I *will* say, however, that we'll need the caching card to ship this change, since otherwise it won't be usable.

jdegoes commented:
Oh, the "informational messages" (such as the MongoDB query plan) will not be moved to card backside. Each card must present information in the way it desires, in a card-specific way. I don't think this can be standardized.

jdegoes commented:
Oh, and can we call "JTable" just "Table". :smile:

cryogenian commented:
<img width="692" alt="2016-03-30 1 33 46" src="https://cloud.githubusercontent.com/assets/10245930/14126164/df549366-f617-11e5-9906-dc2c92aa9d8a.png">
<img width="682" alt="2016-03-30 1 34 13" src="https://cloud.githubusercontent.com/assets/10245930/14126165/df55846a-f617-11e5-96f1-f94e1cdadcb9.png">


+ Added disabled buttons to next action card. 
+ Removed some dead code. 
+ Updated tests. 
+ "JTable" -> "Table" in labels and `aria-label`s

jonsterling commented:
Heh, this next action card stuff looks pretty sweet. nice job, @cryogenian!

cryogenian commented:
@jonsterling please review

jonsterling commented on src/SlamData/Notebook/Editor/Component.purs:
```diff
@@ -333,17 +330,10 @@ peekEvalCell _ _ = pure unit
 peekAnyCell cellId q = do
   for_ (q ^? _NextQuery <<< _Right <<< Next._AddCellType) createCell
-  when (queryShouldRun q) $ runCell cellId
+  runCell cellId
```
I take it that this is to do with the removal of support for caching?

cryogenian replied:
Yes, as far as I know we don't want to suspend evaluation of card now because making view mount is cheap and fast.

jonsterling commented on src/SlamData/Prelude.purs:
```diff
@@ -15,7 +15,8 @@ limitations under the License.
 
 module SlamData.Prelude
-  ( (∘), (⊕)
+  ( (∘), (⊕), (⋙), (⋘), (≡), (≠), (×), (≪), (≫)
```
Very nice. Our of curiosity, what editor are you using, and how do you input these characters? I use vim, and I know how to input a few of them using Vim's "digraph" feature.

cryogenian replied:
I use emacs, with `agda-input` for `purescript-mode` 
They are: \o, \o+, \ggg, \Ll, \==, \neq, \x, \ll, \gg

cryogenian replied:
Oh and for purescript syntax: \to, \l, \all and \:: for `->`, `<-`, `forall` and `::`

jonsterling commented on test/src/Test/Feature.purs:
```diff
@@ -294,28 +302,28 @@ expectNotPresentedWithPropertiesNotRepeatedly properties xPath =
-  -> XPath
-  -> Feature eff o Unit
+  ∷ ∀ eff o
+  . Properties
```
Heh, one of the big benefits of the unicode versions of these things is that all these chars take up just one unit of space! Much less awkward to align, :+1:

jonsterling commented:
I think the code pretty much looks good, though I hope I didn't miss anything important as a result of the unicode diffs.

I am curious about this "caching card"; is there a specification of how it is planned to work? And when do we intend to put it in?

cryogenian commented:
I'll start to work on `Cache/Save` card right after this. We have no specs but I think it should be very similar with `rename/move` dialog.

jonsterling commented:
@cryogenian All right; then, I think I would give this a :+1: to merge, since I think that it is important that everyone be able to base their changes on top of this as soon as possible.

I am a little queasy about going backward in functionality (losing caching) in a PR, but it seems like it should be fine if we just don't do a release until after the caching card is done.

:+1:

cryogenian commented:
Thanks for review!

beckyconning commented:
This is ace, thanks for getting this in @cryogenian.
